modules/library/src/main/scala/zio/elasticsearch/ElasticRequest.scala

```diff
@@ -16,226 +16,244 @@
+    GetByQuery(index = index, query = query, routing = None)
 
-  private[elasticsearch] final case class BulkRequest(
-    requests: List[BulkableRequest],
+  def upsert[A: Schema](index: IndexName, id: DocumentId, doc: A): CreateOrUpdate =
+    CreateOrUpdate(index, id, Document.from(doc), refresh = false, routing = None)
+
+  sealed trait BulkRequest extends ElasticRequest[Unit] with HasRouting[Unit] with HasRefresh[Unit]
+
+  private[elasticsearch] final case class Bulk(
+    requests: List[BulkableRequest[_]],
     index: Option[IndexName],
     refresh: Boolean,
     routing: Option[Routing]
-  ) extends ElasticRequest[Unit, Bulk] {
+  ) extends BulkRequest { self =>
+    def routing(value: Routing): Bulk = self.copy(routing = Some(value))
+
+    def refresh(value: Boolean): Bulk = self.copy(refresh = value)
+
+    def refreshFalse: Bulk = refresh(false)
+
+    def refreshTrue: Bulk = refresh(true)
+
     lazy val body: String = requests.flatMap { r =>
       // We use @unchecked to ignore 'pattern match not exhaustive' error since we guarantee that it will not happen
       // because these are only Bulkable Requests and other matches will not occur.
-      (r.request: @unchecked) match {
-        case CreateRequest(index, document, _, maybeRouting) =>
+      r match {
+        case Create(index, document, _, maybeRouting) =>
           List(getActionAndMeta("create", List(("_index", Some(index)), ("routing", maybeRouting))), document.json)
-        case CreateWithIdRequest(index, id, document, _, maybeRouting) =>
+        case CreateWithId(index, id, document, _, maybeRouting) =>
           List(
             getActionAndMeta("create", List(("_index", Some(index)), ("_id", Some(id)), ("routing", maybeRouting))),
             document.json
           )
-        case CreateOrUpdateRequest(index, id, document, _, maybeRouting) =>
+        case CreateOrUpdate(index, id, document, _, maybeRouting) =>
           List(
             getActionAndMeta("index", List(("_index", Some(index)), ("_id", Some(id)), ("routing", maybeRouting))),
             document.json
           )
-        case DeleteByIdRequest(index, id, _, maybeRouting) =>
+        case DeleteById(index, id, _, maybeRouting) =>
           List(getActionAndMeta("delete", List(("_index", Some(index)), ("_id", Some(id)), ("routing", maybeRouting))))
       }
     }.mkString(start = "", sep = "\n", end = "\n")
   }
 
-  object BulkRequest {
-    def of(requests: BulkableRequest*): BulkRequest =
-      BulkRequest(requests = requests.toList, index = None, refresh = false, routing = None)
+  object Bulk {
+    def of(requests: BulkableRequest[_]*): Bulk =
+      Bulk(requests = requests.toList, index = None, refresh = false, routing = None)
   }
 
-  private[elasticsearch] final case class CreateRequest(
+  sealed trait CreateRequest extends BulkableRequest[DocumentId] with HasRouting[DocumentId] with HasRefresh[DocumentId]
```

Comment:
Let's sort this. And use `HasRefresh` and `HasRouting`. Also, sort methods as well.